File: go/memorysync/checkpoint.go
package memorysync

import (
	"compress/gzip"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// Checkpoint reads the session transcript + memory from the src project dir,
// stores them (raw) + a manifest in the sync-store, pushes, returns the
// checkpoint ID (= the session UUID).
func Checkpoint(uuid, srcCwd, cfgPath string) (string, error) {
	cfg, err := LoadConfig(cfgPath)
	if err != nil {
		return "", fmt.Errorf("load config: %w", err)
	}
	srcProject := ProjectDir(srcCwd)
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("get home dir: %w", err)
	}
	workDir := filepath.Join(home, ".memory-sync", "work", cfg.ProjectID)
	store, err := NewGitStore(cfg.Store.URL, workDir)
	if err != nil {
		return "", fmt.Errorf("new git store: %w", err)
	}

	// build a temp bundle (atomic)
	bundle, err := os.MkdirTemp("", "msync-bundle-")
	if err != nil {
		return "", fmt.Errorf("mktemp bundle: %w", err)
	}
	defer os.RemoveAll(bundle)

	// manifest
	m := Manifest{
		ProjectID:    cfg.ProjectID,
		OriginRoot:   srcCwd,
		OriginDashed: Dashed(srcCwd),
		SessionUUID:  uuid,
		GitHead:      gitHead(srcCwd),
	}

	// transcript: gzip if > 1MB, else copy. A missing transcript is a hard
	// error — without it the checkpoint would be manifest-only and silently
	// lie about restoring a session.
	srcJSONL := filepath.Join(srcProject, uuid+".jsonl")
	fi, err := os.Stat(srcJSONL)
	if err != nil {
		return "", fmt.Errorf("session transcript not found at %s (wrong --cwd or uuid?)", srcJSONL)
	}
	if fi.Size() > 1024*1024 {
		if err := gzipFile(srcJSONL, filepath.Join(bundle, uuid+".jsonl.gz")); err != nil {
			return "", fmt.Errorf("gzip transcript: %w", err)
		}
		m.TranscriptCompressed = true
	} else {
		if err := copyFile(srcJSONL, filepath.Join(bundle, uuid+".jsonl")); err != nil {
			return "", fmt.Errorf("copy transcript: %w", err)
		}
	}
	// memory
	srcMem := filepath.Join(srcProject, "memory")
	if _, err := os.Stat(srcMem); err == nil {
		if err := copyDir(srcMem, filepath.Join(bundle, "memory")); err != nil {
			return "", fmt.Errorf("copy memory: %w", err)
		}
	}
	if err := os.WriteFile(filepath.Join(bundle, "manifest.json"), []byte(Dump(m)), 0644); err != nil {
		return "", fmt.Errorf("write manifest: %w", err)
	}

	if err := store.Put(uuid, bundle); err != nil {
		return "", fmt.Errorf("store put: %w", err)
	}
	return uuid, nil
}

func copyFile(src, dst string) error {
	info, err := os.Stat(src)
	if err != nil {
		return err
	}
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, data, info.Mode())
}

func gzipFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()
	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()
	gz := gzip.NewWriter(out)
	defer gz.Close()
	_, err = io.Copy(gz, in)
	return err
}

func gitHead(cwd string) string {
	cmd := exec.Command("git", "-C", cwd, "rev-parse", "HEAD")
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

File: python/memory_sync/checkpoint.py
"""Checkpoint a session: read transcript+memory, store, write manifest, push."""
import os
import shutil

from . import config as config_mod
from . import encoding
from . import manifest
from . import store as store_mod


def checkpoint(session_uuid: str, src_cwd: str, cfg_path: str) -> str:
    cfg = config_mod.load(cfg_path)
    src_project = encoding.project_dir(src_cwd)
    work_dir = os.path.join(os.path.expanduser("~"), ".memory-sync", "work", cfg.project_id)
    s = store_mod.GitStore(cfg.store_url, work_dir)

    # build a temp bundle dir
    import tempfile
    bundle = tempfile.mkdtemp(prefix="msync-bundle-")
    try:
        # transcript
        src_jsonl = os.path.join(src_project, session_uuid + ".jsonl")
        if os.path.isfile(src_jsonl):
            shutil.copy2(src_jsonl, os.path.join(bundle, session_uuid + ".jsonl"))
        # memory
        src_mem = os.path.join(src_project, "memory")
        if os.path.isdir(src_mem):
            shutil.copytree(src_mem, os.path.join(bundle, "memory"))
        # manifest
        m = manifest.Manifest(
            project_id=cfg.project_id,
            origin_root=src_cwd,
            origin_dashed=encoding.dashed(src_cwd),
            session_uuid=session_uuid,
            git_head=_git_head(src_cwd),
        )
        with open(os.path.join(bundle, "manifest.json"), "w") as f:
            f.write(manifest.dump(m))
        s.put(session_uuid, bundle)
        return session_uuid
    finally:
        shutil.rmtree(bundle, ignore_errors=True)


def _git_head(cwd: str) -> str:
    import subprocess
    try:
        r = subprocess.run(["git", "-C", cwd, "rev-parse", "HEAD"], capture_output=True, text=True)
        return r.stdout.strip() if r.returncode == 0 else ""
    except Exception:
        return ""
